embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/internal/spans/SpansRepository.kt

```diff
@@ -0,0 +1,62 @@
+package io.embrace.android.embracesdk.internal.spans
+
+import io.embrace.android.embracesdk.spans.EmbraceSpan
+import java.util.concurrent.ConcurrentHashMap
+
+/**
+ * Allows the tracking of [EmbraceSpan] instances so that their references can be retrieved with its associated spanId
+ */
+internal class SpansRepository {
+    private val activeSpans: MutableMap<String, EmbraceSpan> = ConcurrentHashMap()
+    private val completedSpans: MutableMap<String, EmbraceSpan> = ConcurrentHashMap()
+
+    /**
+     * Track the [EmbraceSpan] if it has been started and it's not already tracked.
+     */
+    fun started(embraceSpan: EmbraceSpan) {
+        embraceSpan.spanId?.let { spanId ->
+            if (activeSpans[spanId] == null && completedSpans[spanId] == null) {
+                synchronized(activeSpans) {
+                    if (activeSpans[spanId] == null && completedSpans[spanId] == null) {
```

Comment:
nit: this is nested pretty deep. I'd suggest `val spanId = embraceSpan.spanId ?: return`, or possibly creating an abstraction for checking the same condition before & after the synchronized block

Reply:
Yeah I tacked this on pretty late. Let me think about this - early return probably solves this nicely.